### PR TITLE
TimeDistributed with masking

### PR DESCRIPTION
**Pull request by matt-gardner**
This PR fixes two issues with `TimeDistributed`:

1. It passes the mask that's input to `TimeDistributed` on to the wrapped layer, so that wrapped layers can use the mask properly.
2. It allows for layers taking multiple inputs to work directly with `TimeDistributed`.

There are three things to note:

1. It appears you removed the `mask=None` kwarg to the `Layer.call()` method.  That makes handling masking correctly in `TimeDistributed` hard, which is why I added it back in `Dense` and `_Conv`.  Is there a better way to handle this?
2. In `compute_mask`, this implementation only passes through the input mask.  This is clearly not what you want to do in general - you really want to use the wrapped layer's `compute_mask` method, but to get that to work correctly would require much more sweeping changes to the code.
3. I haven't added any tests yet, as I'm not sure what exactly you would like to see tested.  I'm happy to add whatever tests you would like, though.

### CODE REVIEW

File: keras/layers/wrappers.py [hunk 1]
```diff
@@ -97,53 +103,90 @@ def __init__(self, layer, **kwargs):
...
-            inputs = K.reshape(inputs, (-1,) + input_shape[2:])
-            y = self.layer.call(inputs)  # (num_samples * timesteps, ...)
-            # Shape: (num_samples, timesteps, ...)
+            reshaped_xs, reshaped_masks = self.reshape_inputs_and_masks(inputs, mask)
+            outputs = self.layer.call(reshaped_xs, mask=reshaped_masks)
             output_shape = self.compute_output_shape(input_shape)
-            y = K.reshape(y, (-1, input_length) + output_shape[2:])
-
-        # Apply activity regularizer if any:
-        if (hasattr(self.layer, 'activity_regularizer') and
```
Comment (matt-gardner) [inline keras/layers/wrappers.py]: Looks like this was added after I branched off from keras with this layer.  I'll add these lines back in.

File: keras/layers/convolutional.py [hunk 2]
```diff
@@ -141,7 +141,7 @@ def build(self, input_shape):
                                     axes={channel_axis: input_dim})
         self.built = True
 
-    def call(self, inputs):
+    def call(self, inputs, mask=None):
```
Comment (fchollet) [inline keras/layers/convolutional.py]: Not needed.

File: keras/layers/core.py [hunk 3]
```diff
@@ -830,7 +830,7 @@ def build(self, input_shape):
         self.input_spec = InputSpec(min_ndim=2, axes={-1: input_dim})
         self.built = True
 
-    def call(self, inputs):
+    def call(self, inputs, mask=None):
```
Comment (fchollet) [inline keras/layers/core.py]: Not needed.

File: keras/layers/wrappers.py [hunk 4]
```diff
@@ -97,53 +103,90 @@ def __init__(self, layer, **kwargs):
         self.supports_masking = True
 
     def build(self, input_shape):
-        assert len(input_shape) >= 3
-        self.input_spec = InputSpec(shape=input_shape)
-        child_input_shape = (input_shape[0],) + input_shape[2:]
+        if isinstance(input_shape, tuple):
+            input_shape = [input_shape]
+        assert all(len(shape) >= 3 for shape in input_shape), "Need 3 dims to TimeDistribute"
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This should be a `ValueError`.

File: keras/layers/wrappers.py [hunk 5]
```diff
@@ -97,53 +103,90 @@ def __init__(self, layer, **kwargs):
...
+        all_timesteps = [i[1] for i in input_shape]
+        assert len(set(all_timesteps)) == 1, "Tensors must have same number of timesteps"
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This should be a `ValueError`.

File: keras/layers/wrappers.py [hunk 6]
```diff
@@ -97,53 +103,90 @@ def __init__(self, layer, **kwargs):
...
+            reshaped_x = K.reshape(input_i, (-1,) + input_shape[2:])  # (batch_size * timesteps, ...)
+            if mask_i is not None:
+                mask_ndim = K.ndim(mask_i)
+                input_ndim = K.ndim(input_i)
+                if mask_ndim == input_ndim:
+                    mask_shape = input_shape
+                elif mask_ndim == input_ndim - 1:
+                    mask_shape = input_shape[:-1]
+                else:
+                    raise Exception("Mask is of an unexpected shape. Mask's ndim: %s, input's ndim %s" %
```
Comment (fchollet) [inline keras/layers/wrappers.py]: This should be a `ValueError` (never raise `Exception`).

File: keras/layers/wrappers.py [hunk 7]
```diff
@@ -97,53 +103,90 @@ def __init__(self, layer, **kwargs):
...
-        # Apply activity regularizer if any:
-        if (hasattr(self.layer, 'activity_regularizer') and
-           self.layer.activity_regularizer is not None):
-            regularization_loss = self.layer.activity_regularizer(y)
-            self.add_loss(regularization_loss, inputs)
-        return y
+            outputs = K.reshape(outputs, (-1, timesteps) + output_shape[2:])
+        return outputs
+
+    def compute_mask(self, input, input_mask=None):
```
Comment (fchollet) [inline keras/layers/wrappers.py]: Should be `inputs` and `mask`.

**Comment by matt-gardner** [comment]
You mentioned in the contributions requested issue that TimeDistributed still needed work, so I thought it wasn't handling the mask right.  But it appears there's some magic going on in Keras 2 with passing the mask around that I don't understand well enough yet, so I'll just wait on trying to improve things until I've had more time to understand Keras 2.
